params/json.libevm.go

```diff
@@ -0,0 +1,119 @@
+package params
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/libevm/pseudo"
+)
+
+var _ interface {
+	json.Marshaler
+	json.Unmarshaler
+} = (*ChainConfig)(nil)
+
+// chainConfigWithoutMethods avoids infinite recurion into
+// [ChainConfig.UnmarshalJSON].
+type chainConfigWithoutMethods ChainConfig
+
+// chainConfigWithExportedExtra supports JSON (un)marshalling of a [ChainConfig]
+// while exposing the `extra` field as the "extra" JSON key.
+type chainConfigWithExportedExtra struct {
+	*chainConfigWithoutMethods              // embedded to achieve regular JSON unmarshalling
+	Extra                      *pseudo.Type `json:"extra"` // `c.extra` is otherwise unexported
+}
+
+// UnmarshalJSON implements the [json.Unmarshaler] interface.
+func (c *ChainConfig) UnmarshalJSON(data []byte) error {
+	extras := registeredExtras
+
+	if extras != nil && !extras.reuseJSONRoot {
+		return c.unmarshalJSONWithExtra(data)
+	}
+
+	if err := json.Unmarshal(data, (*chainConfigWithoutMethods)(c)); err != nil {
+		return err
+	}
+	if extras == nil {
+		return nil
+	}
+
+	// Invariants if here:
+	// - reg.reuseJSONRoot == true
+	// - Non-extra ChainConfig fields already unmarshalled
+
+	c.extra = extras.chainConfig.NilPointer()
+	if err := json.Unmarshal(data, c.extra); err != nil {
+		c.extra = nil
+		return err
+	}
+	return nil
+}
+
+// unmarshalJSONWithExtra unmarshals JSON under the assumption that the
+// registered [Extras] payload is in the JSON "extra" key. All other
+// unmarshalling is performed as if no [Extras] were registered.
+func (c *ChainConfig) unmarshalJSONWithExtra(data []byte) error {
+	cc := &chainConfigWithExportedExtra{
+		chainConfigWithoutMethods: (*chainConfigWithoutMethods)(c),
+		Extra:                     registeredExtras.chainConfig.NilPointer(),
+	}
+	if err := json.Unmarshal(data, cc); err != nil {
+		return err
+	}
+	c.extra = cc.Extra
+	return nil
+}
+
+// MarshalJSON implements the [json.Marshaler] interface.
+func (c *ChainConfig) MarshalJSON() ([]byte, error) {
```

Comment:
I noticed in a couple cases that `MarshalJSON()` being defined on the value pointer makes a difference.
Since it is only a couple instances so far I switched the uses to `&` to move on, but perhaps this should be defined on the value to avoid invoking the default MarshalJSON when the receiver is non-pointer.

Reply:
Interesting! Do you remember where this was, so I can take a look before I merge this? I defined it on the pointer because that's what tends to be carried around, e.g. in `vm.EVM`.

Reply:
I'm going to leave it as defined on the pointer for now. In the geth code base there are [only two instances](https://github.com/search?q=repo%3Aethereum%2Fgo-ethereum+%2F%5B%5E*%26%5Dparams%5C.ChainConfig%2F&type=code) where a non-pointer `ChainConfig` is used (note that most of the results are either `new(ChainConfig)` or immediately address the value with `&`).

This is in contrast to the [65 files with a `*ChainConfig`](https://github.com/search?q=repo%3Aethereum%2Fgo-ethereum+*params.ChainConfig&type=code).

Reply:
The *ChainConfig instances will still work if the receiver is value. We can leave it as is for now.